### ISSUE TITLE
(Draft) RFC: Gates for Events

### ISSUE THREAD

**Issue by agustif** [calcom/cal.com#1913]
After recently joining the team at cal.com, I’ve been tasked re-factoring our current implementation of the web3 integration of cryptowallets with metamask.

As I started porting the integration to use wagmi library. We saw the potential for a new abstraction to fit both the new web3 usability (Booking restriction of holding an ERC-20 Token by introducing the SmartContractAddreess in the event additional settings)

We propose the implementation of a new Gate or gating system, which could include some different Gates to use in the core, and also by extended by plugins or 3rd party integrations by providing their owng gates when appropiate

In this new system a customer could set in each booking type could select one or different **Gates** enabled on it, with different settings or verification checks in place.

**Web3 Gate**

As an example, the new implementation of the web3 gate could be enabled, and have some settings.

By **default** and as **only enabled** and not other settings, you could require the user has a valid crypto wallet and validates with it before booking.

By adding a **smartContractAddress**, we could check if the user has a balance bigger than 0 of the specific token.

By adding a **minAmount**, you could limit the booking of holders of more than X token.

By adding a **minHodlingTime** amount, you could check if x hours/days in the past, they were also holding the same minAmount of such token.

**Whitelist wallets**

we could provide gating to only an allowed list of wallets, good for KYC pre-approved wallets and such.

**Password Gate**

If enabled, would require a password input, which the user booking should provide before being able to book the event.

**Whitelist** **Email Gate**

This one could be useful for companies.

Allow only bookings by a whitelisted of either full emails, or domains. Maybe an easy upload or a contacts view will be required. A contact could both 

**Disposable Email Gate**

Implementing an npm package like 

Deep Email Validator [https://www.npmjs.com/package/deep-email-validator](https://www.npmjs.com/package/deep-email-validator)

Allow the customer to set the event type to either disallow or warn you about the use of disposable emails by the users booking an event.

**Ip/geolocation Gate**

This one hasn’t been asked for yet as far as I know. But for big enterprise 

**AnonGate**

Allow hiding different stuff from the booking like the link #1893

Allow the generation of one or several random or not but different than the canonical aliases to the booking. That could have an expiryDate or not. We could also allow tracking so you can have different links in different platforms/marketing channels or somehing like that… and know how many visits/bookings/got you each.

**Comment by PeerRich** [comment]
love this. i thinkg "Disposable Email Gate" should be a default core feature and always warn booking receivers about the incoming booking request and not be a "gate"

**Comment by zomars** [comment]
Would the paid bookings fit into these gates? We could do abstract away into a StripeGate, PayPalGate, etc.

**Comment by agustif** [comment]
> Would the paid bookings fit into these gates? We could do abstract away into a StripeGate, PayPalGate, etc.

Yeah I think so! I think each integration could provide it's extra gates when enabled.

In another note We also commented of subscription roles in stripe or patreon to book events. (Patreon is available in next-auth as SSO provider I think) and seems like a good perk for patrons exclusive booking access.
